docs/cookbook/rag.md

```diff
@@ -1,7 +1,424 @@
+
+- query the dataset with a topic of our choosing
+- retrieve a number of relevant articles
+- ask our chat model to summarize the relevant ones
+- show how to query the dataset both locally and using a vector database [(Pinecone)](https://docs.pinecone.io/)
+- show how Mirascope can simplify RAG.
+
+!!! note
+
+    The following code snippets have been moved around for the sake of clarity in the walkthrough, and may not work if you copy and paste them. For a fully functional script, take a look at the [code](https://github.com/Mirascope/mirascope/blob/main/cookbook/rag_examples/) in our repo.
+
+## Before we get started
+
+If you haven't already, it will be worth taking a look some of our relevant concept pages for a more detailed explanation of the functionality we'll be using in this walkthrough.
+
+- [Pydantic Prompts](https://docs.mirascope.io/latest/concepts/pydantic_prompts/) (template variables, messages)
+- [LLM Convenience Wrappers](https://docs.mirascope.io/latest/concepts/llm_convenience_wrappers/#openaichat)
+
+In addition, here are the variables we will be using in this cookbook recipe:
+
+```python
+# config.py
+
+from typing import Optional
+
+from pydantic_settings import BaseSettings, SettingsConfigDict
+
+MODEL = "gpt-3.5-turbo"
+EMBEDDINGS_MODEL = "text-embedding-ada-002"
+MAX_TOKENS = 1000
+TEXT_COLUMN = "Text"
+EMBEDDINGS_COLUMN = "embeddings"
+FILENAME = "news_article_dataset.pkl"
+URL = "https://raw.githubusercontent.com/Dawit-1621/BBC-News-Classification/main/Data/BBC%20News%20Test.csv"
+PINECONE_INDEX = "news-articles"
+PINECONE_NAMESPACE = "articles"
+
+class Settings(BaseSettings):
+    openai_api_key: Optional[str] = None
+    pinecone_api_key: Optional[str] = None
+
+    model_config = SettingsConfigDict(env_file=".env")
+```
+
+```python
+# .env
+
+OPENAI_API_KEY = "YOUR_OPENAI_API_KEY"
+PINECONE_API_KEY = "YOUR_PINECONE_API_KEY"
+```
+
+## Load and preprocess the data
+
+Before we load raw data into a pandas `Dataframe`, we have to handle token limits for both chat and embeddings models. In our case, where `gpt-3.5-turbo` has a lower token limit (4096) than the `text-embedding-ada-002` (8191), the chat becomes the limiting factor. A crude solution is to chunk each article into equal token lengths of at most `MAX_TOKENS=1000`, so we may fit 3 article snippets as well as any text in our hand-written section of the prompt. The function `split_text()` below contains the actual chunking logic.
+
+For counting the token lengths, we will use [tiktoken](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) since we will be using OpenAI for both the chat and embedding models. [tiktoken](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb) is a useful library provided by OpenAI for encoding strings and decoding tokens for their models.
```

Comment:
"for counting the token lenghts" -> "For counting the number of tokens in an article"